src/i2cint.rs

```diff
@@ -0,0 +1,386 @@
+//! Inter-Integrated Circuit (I2C) bus interruption
+
+use crate::i2c::{I2c, Instance, SclPin, SdaPin};
+use crate::rcc::{Clocks, APB1};
+use crate::time::Hertz;
+use core::fmt;
+
+/// I2c errors.
+#[derive(Debug, Copy, Clone)]
+pub enum I2cError {
+    /// Device is busy, can't start something else
+    DeviceBusy,
+    /// Received a nack
+    Nack,
+    /// Error happened on the bus
+    BusError,
+    /// Arbitration loss,
+    ArbitrationLoss,
+    /// Overrun detected (salve mode)
+    Overrun,
+    /// Unable to compute the stop state because previous state was not expected
+    StateError,
+    /// Transfer complete status but nothing do do next
+    TransferCompleteNoRead,
+    /// Send buffer is empty
+    TxBufferEmpty,
+}
+
+/// State of i2c communication.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum State {
+    /// I2c is Idle
+    Idle,
+    /// Send has started
+    TxStart,
+    /// Ready for send
+    TxReady,
+    /// Data written in send buffer
+    TxSent,
+    /// Send is complete, but device is not stopped
+    TxComplete,
+    /// Send is complete
+    TxStop,
+    /// Receive has started
+    RxStart,
+    /// Ready for receive
+    RxReady,
+    /// Received is complete
+    RxStop,
+    /// Nack for send
+    TxNack,
+}
+
+impl core::fmt::Display for State {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// I2c1Int provides interface to communicate with i2c devices using interruptions.
+pub struct I2cInt<I2C, PINS> {
```

Comment:
A `free` function is missing, so that the user can regain access to the pins, if I2c is no longer needed:

https://github.com/stm32-rs/stm32f3xx-hal/blob/9e29ec6b1dba461eb1001ea6eee2bcf255fda59f/src/i2c.rs#L201-L204